### PR TITLE
Fix Stronghold boxes

### LINKED ISSUE #4300: Stronghold boxes are bad
image

### PR DESCRIPTION
**Pull request by Arch-Shaman**
Fixes #4300

Big Teams (5v5+)
image

Small Teams (3v3 - 5v5)
image

1v1 - 2v2
image

Also tries to add ffa boxes
image

(Supports both 4 way and 8 way)

### CODE REVIEW

File: LuaRules/Configs/StartBoxes/Stronghold_beta.lua [hunk 1]
```diff
@@ -1,32 +1,919 @@
...
+end
+
+-- check for FFA
+local isFFA = Spring.Utilities.Gametype.isFFA()
+
+if not isFFA then -- give extra boxes in big teams/coop
+	local count1 = #Spring.GetTeamList(1)
+	local count2 = #Spring.GetTeamList(2)
+	local players = math.max(count1, count2)
+	if players > 4 then
```
Comment (sprunk) [inline LuaRules/Configs/StartBoxes/Stronghold_beta.lua:890]: Why not use `Spring.Utilities.Gametype.isSmallTeams` (and `isBigTeams`)?

File: LuaRules/Configs/StartBoxes/Stronghold_beta.lua [hunk 2]
```diff
@@ -1,32 +1,919 @@
...
+		boxes[2].boxes[2] = team2flexbox
+	end
+else -- add boxes to allyteam 3 and 4.
+	boxes[3] = {startpoints = {}, nameLong = names[3], nameShort = shorts[3], boxes = {}}
+	boxes[4] = {startpoints = {}, nameLong = names[4], nameShort = shorts[4], boxes = {}}
+	boxes[3].boxes[1] = team3box
+	boxes[4].boxes[1] = team4box
+	boxes[3].startpoints = team3boxstart
+	boxes[4].startpoints = team4boxstart
+	return boxes, {4}
```
Comment (sprunk) [inline LuaRules/Configs/StartBoxes/Stronghold_beta.lua:916]: What about 8-way?

File: LuaRules/Configs/StartBoxes/Stronghold_beta.lua [hunk 3]
```diff
@@ -1,32 +1,919 @@
...
+names[2] = "Southeast"
+names[3] = "Northeast"
+names[4] = "Southwest"
+
+local shorts = {}
+shorts[1] = "NW"
+shorts[2] = "SE"
+shorts[3] = "NE"
+shorts[4] = "SW"
+
```
Comment (sprunk) [inline LuaRules/Configs/StartBoxes/Stronghold_beta.lua:832]: The big teams version should probably be just "east" and "west".

**Review by sprunk — CHANGES_REQUESTED** [review]
Boxes are indexed from 0

File: LuaRules/Configs/StartBoxes/Stronghold_beta.lua [hunk 4]
```diff
@@ -882,38 +1131,88 @@ end
...
+		else
+			local team1 = MergeStarts(team1boxstart, team7boxstart)
+			local team2 = MergeStarts(team2boxstart, team8boxstart)
+			boxes[1].startpoints = team1
+			boxes[2].startpoints = team2
+			boxes[1].boxes[2] = team7box
+			boxes[2].boxes[2] = team8box
+		end
+	end
+else -- 8 way FFA.
```
Comment (sprunk) [inline LuaRules/Configs/StartBoxes/Stronghold_beta.lua:1198]: If you return all 8 boxes for any FFA, the 4-way risks using some corner boxes and some with the geos. While boxes don't need to be fair if we're cramming more people than there are equal start positions, they still should if we can afford it.

Make 4-way either use the corners or the geos (doesn't matter which, can even be random)

Comment (Arch-Shaman) [reply LuaRules/Configs/StartBoxes/Stronghold_beta.lua:1198]: alright on it.

Comment (Arch-Shaman) [reply LuaRules/Configs/StartBoxes/Stronghold_beta.lua:1207]: 3476e9a (resolves?)

**Comment by Arch-Shaman** [comment]
> Boxes are indexed from 0

Is this still applicable? Do I need to redo everything with index 0?

File: LuaRules/Configs/StartBoxes/Stronghold_beta.lua [hunk 5]
```diff
@@ -1,32 +1,1247 @@
...
+local teamcount = 0
+do
+	local allyteams = Spring.GetAllyTeamList()
+	for a = 1, #allyteams - 1 do -- Presumes last allyteamID is gaia.
+		local allyid = allyteams[a]
+		if #Spring.GetTeamList(allyid) > 0 then
+			teamcount = teamcount + 1
+		end
+	end
+end
```
Comment (sprunk) [inline LuaRules/Configs/StartBoxes/Stronghold_beta.lua:1144]:
```suggestion
local teamcount = Spring.Utilities.GetTeamCount()
```

File: LuaRules/Configs/StartBoxes/Stronghold_beta.lua [hunk 6]
```diff
@@ -1,32 +1,1247 @@
...
+		return boxes, {8}
+	else
+		
+		if r > 5 then
+			boxes[2].boxes[1] = team3box
+			boxes[3].boxes[1] = team4box
+			boxes[2].startpoints = team3boxstart
+			boxes[3].startpoints = team4boxstart
+		else
+			boxes[0].boxes[1] = team5box
```
Comment (sprunk) [inline LuaRules/Configs/StartBoxes/Stronghold_beta.lua:1234]: seems to have incorrect names?

Comment (Arch-Shaman) [reply LuaRules/Configs/StartBoxes/Stronghold_beta.lua:1225]: You're right. Good catch. Should be i + 1

**Comment by Arch-Shaman** [comment]
I did something wrong somewhere..

**Comment by sprunk** [comment]
I expect to have the time to actually test this after 04-07.

**Comment by Arch-Shaman** [comment]
Error: [LuaRules::RunCallInTraceback] error=2 (LUA_ERRRUN) callin=LoadCode trace=[Internal Lua error: Call failure] [LuaVFS::Include(synced=1)][pcall] file=LuaRules/gadgets.lua error=2 ([LuaVFS::Include(synced=1)][pcall] file=LuaRules/Configs/StartBoxes/Stronghold V4.lua error=2 ([LuaVFS::Include(synced=1)][pcall] file=LuaRules/Configs/StartBoxes/Stronghold_beta.lua error=2 ([string "LuaRules/Configs/StartBoxes/Stronghold_beta..."]:1119: attempt to index field '?' (a nil value)) ptop=1 cenv=0) ptop=2 cenv=1) ptop=3 cenv=0
stack traceback:
    [C]: in function 'Include'
    [string "LuaRules/draw.lua"]:15: in main chunk

**Comment by Arch-Shaman** [comment]
I didn't reindex the initiator part.

**Pull request merged**